Improve LIP0039 to align it to the implementation (#431)
LiskHQ/lips: proposals/lip-0039.md

    *   The Prover generates `siblingHashes` by merging together the sibling hashes of the single paths in the correct order of application, and transmits the complete proof to the Verifier. The complete proof contains a common array `siblingHashes` and an array `queries`, containing a query object for each key in `queryKeys`, respecting the same order.
*   Verifier: 
    *   The Verifier uses `siblingHashes` to recalculate the Merkle root starting from the leaf nodes derived from `queryKeys` and `queries`, and checks that it equals `merkleRoot`. 
    *   By comparing the original array of `queryKeys` with the keys in the `proof.queries` array, the Verifier checks that all keys in `queryKeys` are present in the tree.

As shown, a non-inclusion proof is generated by providing a proof of inclusion for another leaf or a default empty node in the path of the target leaf node, indicating that there are no other leaves in the subtree (see Figure 2). The protocol for a non-inclusion proof is thus exactly the same as for an inclusion proof, except that in the last step, where:

*   Verifier: 

>     *   By comparing the original array of `queryKeys` with the keys in the `proof.queries` array, the Verifier checks that no keys in `queryKeys` are present in the tree.

Author: I would limit cosmetic changes to the current update

### Node Batching Optimization

One of the bottlenecks for a fast update of the sparse Merkle tree is the I/O bandwidth to retrieve the nodes from the database. To reduce the number of database accesses, we introduce the following optimization: Instead of storing every node individually, we store all nodes belonging to the same subtree of height `h` 'batched' together; furthermore, only the subtree external nodes are stored, while the internal nodes are computed on the fly whenever the subtree is updated. This allows to reduce the number of databse accesses by a factor `2^h` at the cost of a slightly more complex and computationally expensive update protocol.

> In the following section however, we do not implement this optimization and rather focus on the clarity of the specifications.

Reviewer: ```suggestion
In the following section, however, we specify sparse Merkle trees without node batching in order to focus on the clarity of the specifications.
```

| Name             | Type    | Value       |
| ---------------- |---------| ------------|
| EMPTY_HASH       | bytes   | SHA-256("") |
| LEAF_PREFIX      | bytes   | ascii encoding of "LSK_SMTL_"  |
| BRANCH_PREFIX    | bytes   | ascii encoding of "LSK_SMTB_"  |
| KEY_LENGTH_BYTES | integer | Length of keys in bytes. Set when initializing the data structure |

> Each node in the tree has a `data` and a `hash` property. Leaf nodes store the key and value in their data property, while branch nodes store the hashes of the child nodes. Default empty nodes have a constant hash `EMPTY_HASH`.  In summary, we define 3 different node types:

Author: fixed

        return EMPTY_HASH
```

> Here the function `sha256` returns the SHA-256 hash of the input. Similar to what we do for regular Merkle trees, we use a different hash function for leaf and branch nodes to protect against second preimage attacks and prefix the hashed data with different constants.

Reviewer 2: `sha256` -> `hash`

Author: I use sha256 to distinguish the sha hashing function from the hash property of the node

1. Each subtree with exactly one non-empty leaf is replaced by the leaf itself.
2. Each subtree containing only empty nodes is replaced by a constant node with hash value equal to `EMPTY_HASH`.

### Root Hash Calculation

The Merkle root of a dataset is computed as follows:

1. The Merkle root of an empty dataset is set to the constant value `EMPTY_HASH`.
2. The Merkle root of a dataset with a single element is set to the leaf hash of that element.
3. Otherwise, the Merkle root is the root of the tree resulting from the update protocol outlined here, where `root` is the current root of the tree and `keys` and `values` are the keys and corresponding values to be inserted.

```python

> def update(

Author: (c.c. @shuse2) To align with the implementation, I would leave the name `update` here and actually call the `remove` function if the passed value is empty.

> def update(

Reviewer: That I would find rather more confusing. If we want to keep the name, I would still separate between `update` and `remove`.

Should we then check, however, that none of the values is empty?

> def update(

Reviewer 3: i think implementation wise, i would prefer not to change it now to separate update/remove coz it's quite complex to detach them now

> def update(

Reviewer: For the LIP I would keep remove and update still separate as this is how the LIP was before and it is the cleaner solution.

> def update(

Reviewer: As discussed, we keep the functions as is and I resolve this comment.

    root: bytes,
    keys: list[bytes],
    values: list[bytes],
) -> LeafNode | BranchNode | EmptyNode:

    assert len(keys) == len(values)

    if len(keys) == 0:

>         return root

Author: fixed

            return LeafNode(keys[0], values[0])

    # We split the keys into two lists, depending on whether the key bit at index 
    # equal to height is 0 or 1.
    idx = split_index(keys, height)
    left_keys = keys[:idx] 
    left_values = values[:idx]
    right_keys = keys[idx:] 
    right_values = values[idx:]

    # We calculate the left and right child recursively.
    if isinstance(current_node, EmptyNode):
        left_node = _update(left_keys, left_values, EmptyNode(), height + 1)
        right_node = _update(right_keys, right_values, EmptyNode(), height + 1)
    elif isinstance(current_node, LeafNode):
        if is_bit_set(current_node.key, height):
            left_node = _update(left_keys, left_values, EmptyNode(), height + 1)
            right_node = _update(right_keys, right_values, current_node, height + 1)
        else:
            left_node = _update(left_keys, left_values, current_node, height + 1)
            right_node = _update(right_keys, right_values, EmptyNode(), height + 1)
    elif isinstance(current_node, BranchNode):
        left_node = get_node(current_node.left_hash)

>         right_node = get_node(current_node.right_hash)

Author: The `get_node` function was introduced after the `_update` function, I know moved the definition to the top of the update section

>         right_node = get_node(current_node.right_hash)

Author: Regarding the second point, I define the `get_node` function to receive an hash value and return a Node instance, so I think these specs are correct.

>         right_node = get_node(current_node.right_hash)

Author: Regarding the second comment, you mean rename `left_node` -> `current_left_node`? I don't find that very clear, maybe `left_child` would be clearer

>         right_node = get_node(current_node.right_hash)

Reviewer: Regarding the first comment it looks all good.

Regarding the second comment, my proposal was to not use `leftChildNode` and `rightChildNode` both as temporary value for the current children as well as the value for the future children. Concretely, instead do:

```
    currentLeftChild = getNode(currentNode.leftHash)
    currentRightChild = getNode(currentNode.rightHash)
    leftChildNode = updateNode(leftKeys, leftValues, currentLeftChild, height + 1)
    rightChildNode = updateNode(rightKeys, rightValues, currentRightChild, height + 1)
```

Or, altogether avoid the temporary variables.

>         right_node = get_node(current_node.right_hash)

Author: done

        right_node = _update(right_keys, right_values, right_node, height + 1)

    return BranchNode(left_node.hash, right_node.hash)
```

Here the function `get_node` returns the node corresponding to the given hash value, for instance by fetching it from a database.
We use the following utility functions:

```python
def split_index(keys: list[bytes], height: int) -> int:
    '''
    This function returns the index of the first key with bit at index height set to 1.
    '''
    for idx, key in enumerate(keys):
        if is_bit_set(key, height):
            return idx
    return len(keys)
```

```python
def is_bit_set(bits: bytes, i: int) -> bool:
    '''
    This function returns true if the i-th bit of bits is 1, false otherwise.
    '''

    shifted = bits[i // 8] << (i % 8)
    BIT_COMP = int.from_bytes(b"\x80", "big")

    return (shifted & BIT_COMP) == BIT_COMP
```

### Removing a Leaf Node

A certain key-value pair can be removed from the tree by deleting the corresponding leaf node and rearranging the affected nodes in the tree. The following protocol can be used to remove a key `key` from the tree.

> ```python

Author: fixed and fixed some invalid typing

def remove(root: bytes, key: bytes) -> LeafNode | BranchNode | EmptyNode:
    ancestor_nodes: list[BranchNode] = []
    current_node = root
    h = 0
    current_node_sibling = EmptyNode()

    # Collect all ancestor nodes through traversing the binary expansion by height
    # End of the loop ancestor_nodes has all the branch nodes
    # current_node will be the leaf/node we are looking to remove
    while isinstance(current_node, BranchNode):
        ancestor_nodes.append(current_node)
        if is_bit_set(key, h):
            current_node_sibling = get_node(current_node.left_hash)
            current_node = get_node(current_node.right_hash)
        else:
            current_node_sibling = get_node(current_node.right_hash)
            current_node = get_node(current_node.left_hash)
        h += 1

    # When current_node is empty, nothing to remove
    if isinstance(current_node, EmptyNode):
        return current_node

    # When the input key does not match node key, nothing to remove
    if current_node.key != key:
        return current_node

    bottom_node = EmptyNode()

    if isinstance(current_node_sibling, LeafNode):

>         # current_node has a leaf sibling,

Author: You mean that we should update the comment?

           {key:a9, value:9e8e8c37, bitmap:07}]}.
```

*Given these keys and values, the hash of the node can be recomputed. For example, to the leaf node with <code>key=00110011</code> corresponds the hash <code>leafHash(33 || 4e07408562bedb8b60ce05c1decfe3ad16b72230967de01f640b7e4729b49fce) = 00be9f2ec46f47e14965f0cb9903f09bc6fe30244109c7c5310180a2251c75cc</code>.*

#### Proof Construction

The properties and protocol for an inclusion proof have been introduced in the Rationale section, above. In this section, we specify how these properties are calculated. We assume that the queried keys have a fixed length `KEY_LENGTH_BYTES`.

We define the following data structures:

> 

Author: Do you have a proposal for the name change?

> 

Reviewer: `QueryProof` -> `ExtendedProof`
Otherwise, as discussed, we try to follow the implementation.

> 

Author: went for extended proof

```python
@dataclass
class Query:
    key: bytes
    value: bytes
    bitmap: bytes

    def __post_init__(self) -> None:
        # The binary bitmap is the binary string obtained from the bitmap.
        # We will use it also to keep track of the nodes that have been visited
        # and the current height of the query.
        self.binary_bitmap = binary_expansion(self.bitmap).lstrip("0")
    
    @property
    def height(self) -> int:
        return len(self.binary_bitmap)

    @property
    def binary_path(self) -> str:
        # Convert the key to binary string and take the first self.height digits.
        return binary_expansion(self.key)[:self.height]
```

```python
@dataclass
class QueryProof(Query):
    ancestor_hashes: list[bytes]
    sibling_hashes: list[bytes]
```

```python
@dataclass
class Proof:

>     sibling_hashes: list[bytes]

Author: now everything is camelCase

    queries: list[Query]
```

We make use of the following utility functions:

```python
def binary_expansion(key: bytes):
    '''
    Convert a key to its binary representation
    '''
    return f"{int.from_bytes(key, 'big'):0{8*len(key)}b}"
```

```python
def are_sibling_queries(q1: Query, q2: Query) -> bool:
    '''
    This function returns true if two queries are sibling,
    i.e. they point to two sibling nodes, false otherwise.
    '''
    if len(q1.binary_bitmap) != len(q2.binary_bitmap):
        return False

>     if q1.binary_key[:q1.height - 1] != q2.binary_key[:q.height - 1]:

Author: fixed

        queries.insert(insert_index, q)

    return queries
```

The following function generates the query response to a single key, including the sibling hashes and the hash of the visited nodes.

```python
def generate_query_proof(query_key: bytes, root: bytes) -> QueryProof:
    ancestor_hashes = []
    sibling_hashes = []
    current_node = get_node(root)
    binary_bitmap = ""

    # Starting from the root, we navigate down the tree.
    h = 0
    while isinstance(current_node, BranchNode):
        ancestor_hashes.append(current_node.hash)
        if is_bit_set(query_key, h):
            sibling_hash = current_node.left_hash
            current_node = get_node(current_node.right_hash)
        else:
            sibling_hash = current_node.right_hash
            current_node = get_node(current_node.left_hash)

        if sibling_hash == EmptyNode.hash:
            binary_bitmap = "0" + binary_bitmap
        else:
            binary_bitmap = "1" + binary_bitmap
            sibling_hashes.append(sibling_hash)
        h += 1

    if isinstance(current_node, EmptyNode):
        # query_key is not part of the tree
        return QueryProof(

>             query_key, EMPTY_VALUE, binary_to_bytes(binary_bitmap), ancestor_hashes, sibling_hashes

Reviewer: I think the function `binary_to_bytes` was not introduced.

Author: I just calculate the conversion explicitly

        )
    # Else current_node is a leaf node
    assert isinstance(current_node, LeafNode)
    ancestor_hashes.append(current_node.hash)
    return QueryProof(
        current_node.key, current_node.value, binary_to_bytes(binary_bitmap), ancestor_hashes, sibling_hashes
    )
```

Note that the first digit of a query bitmap is always a 1, else the tree would have been invalidly constructed (a leaf node can not have an empty node as sibling). As a consequence, the hex encoded value of the query bitmap corresponds to a unique series of digits. 

To generate a complete inclusion proof, several query responses are combined together, and their sibling hashes are merged in a single array.
Given an array of keys `query_keys`, the inclusion proof is obtained from the function `generate_proof`.

```python

> def generate_proof(root: bytes, query_keys: list[bytes]) -> Proof:

Author: Why do you think this is required?

    if len(query_keys) == 0:
        return Proof([], [])

>     query_proofs = [await generate_query_proof(k, root, 0) for k in query_keys]

Author: fixed

        if q.binary_bitmap[0] == "1":
            node_hash = q.sibling_hashes.pop()
            if not node_hash in sibling_hashes and not node_hash in ancestor_hashes:
                sibling_hashes.append(node_hash)

        q.binary_bitmap = q.binary_bitmap[1:] # Modifying the binary bitmap also changes the height.
        query_proofs = insert_query(q, query_proofs)

    return Proof(sibling_hashes, queries)
```

#### Proof Verification

To verify the validity of an (non-)inclusion proof, the general strategy is to recalculate the Merkle root starting from the provided proof. On top of this, the Verifier checks that the proof queries corresponds to a (non-)inclusion proof by comparing them to the query keys that were used to generate the proof.

We define a function to verify inclusion proofs, `verify_inclusion_proof`, and a function to verify non-inclusion proofs, `verify_non_inclusion_proof`. Both function just call the `verify` function after checking if the query key has a corresponding proof query giving a (non-)inclusion proof using the `is_inclusion_proof` function.

```python
def is_inclusion_proof(query_key: bytes, query: Query) -> bool:

>     return query_key == query.key and query.value != EMPTY_VALUE

Reviewer: `EMPTY_VALUE` is not defined

            return False

    return verify(query_keys, proof, root)
```

```python
def verify_non_inclusion_proof(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
    # Check that there is a query for each key.
    if len(query_keys) != len(proof.queries):
        return False

    for key, query in zip(query_keys, proof.queries):
        # This is the only difference between an inclusion and a non-inclusion proof.
        if is_inclusion_proof(key, query) == True:
            return False
            
    return verify(query_keys, proof, root)
```

```python
def verify(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
    for key, query in zip(query_keys, proof.queries):
        # The bitmap does not contain extra leading 0s.
        if query.bitmap[0] == 0:
            return False
        if len(key) != KEY_LENGTH_BYTES:
            return False

        # q is an inclusion proof for k or a default empty node.
        if key == query.key:
            continue

        # q is an inclusion proof for another leaf node
        common_prefix = os.path.commonprefix([binary_expansion(query.key), binary_expansion(key)])

>         if len(query.binary_bitmap) > len(common_prefix):

Reviewer: What would be a tree that shows that this condition is necessary? Maybe we could also discuss in person.

            # q does not give an non-inclusion proof for k.
            return False

    filtered_queries = []

>     filter: dict[bytes, bool] = {}

Reviewer: I would not use this additionally data structure and instead make `filtered_queries` a dictionary `q.binary_paty -> q`.

Author: filtered_queries needs to be an array for calculate root, and the order is important

    duplicate_queries = {}
    for q in proof.queries:
        # Remove duplicate queries preserving order.
        # This can happen if the same query is given for different query_keys, as for inclusion proofs or non-inclusion proofs pointing to a leaf node
        # or for non-inclusion proofs pointing to the same empty node.
        # To do this, we check the binary path (key binary expansion up to the query height)

>         if q.binary_path not in filter:

Reviewer: I don't understand why unique q.binary_path and not unique q.key. Maybe we can discuss in person.

Author: updated after discussion, now we check that duplicate queries (queries with equal binary path) point to exactly the same node

Author: @shuse2 We updated the LIP to fix a possible vulnerability. Can you double check how this is done in the DB?

Reviewer 3: the verification in the implementation is done [here](https://github.com/LiskHQ/lisk-db/blob/main/src/sparse_merkle_tree/smt.rs#L845-L877), which is used in https://github.com/LiskHQ/lisk-db/blob/main/src/sparse_merkle_tree/smt.rs#L1528

Reviewer: Can be resolved from my side.

            filtered_queries.append(q)
            filter[q.binary_path] = True

        # Check that if a key appears in several queries, than these queries are exactly the same.
        if q.key not in duplicate_queries:
            duplicate_queries[q.key] = q
        else:
            duplicate_query = duplicate_queries[q.key]
            if q.key != duplicate_query.key or q.bitmap != duplicate_query.bitmap or q.value != duplicate_query.value:
                return False

    # Finally calculate the root from the provided proof and check that it matches the input Merkle root.

>     return calculate_root(proof.sibling_hashes, filtered_queries) == root

Author: now everything is camelCase

```

```python
def calculate_root(sibling_hashes: list[bytes], queries: list[Query]) -> bytes:
    sorted_queries = sorted(queries, key=lambda q: (-q.height, q.key))

    while len(sorted_queries) > 0:
        q = sorted_queries.pop(0)

        # We reached the top of the tree, return merkle root
        if q.height == 0:
            # Check that no extra bits where appended to the bitmap.
            assert len(q.binary_bitmap) == 0
            assert len(sibling_hashes) == 0
            return q.hash

        # we distinguish three cases for the sibling hash:
        # 1. sibling is next element of sorted_queries
        if len(sorted_queries) > 0 and are_sibling_queries(q, sorted_queries[0]):

>             sibling_hash = sorted_queries[0].hash

Reviewer: What is the hash of a query, it is not a property of the `Query` class.

Author: fixed

            # We are merging two branches.
            # Check that the bitmap at the merging point is consistent with the nodes type.
            if sorted_queries[0].hash == EMPTY_HASH:
                assert q.binary_bitmap[0] == "0"
            else:
                assert q.binary_bitmap[0] == "1"
            if q.hash == EMPTY_HASH:
                assert sorted_queries[0].binary_bitmap[0] == "0"
            else:
                assert sorted_queries[0].binary_bitmap[0] == "1"

            # Check that the bitmap coincide from the merging point up.

>             assert len(q.binary_bitmap) == len(sorted_queries[0].binary_bitmap)

Reviewer: So the length of the `binary_bitmap` is already checked as `are_sibling_queries` returned `True`. 

```
def are_sibling_queries(q1: Query, q2: Query) -> bool:
    if len(q1.binary_bitmap) != len(q2.binary_bitmap):
        return False

    if q1.binary_key[:q1.height - 1] != q2.binary_key[:q.height - 1]:
        return False

    return q1.binary_key[q1.height - 1] != q.binary_key[q1.height - 1]

```

Author: removed unnecessary check

           {key:a9, value:9e8e8c37, bitmap:07}]}.
```

*Given these keys and values, the hash of the node can be recomputed. For example, to the leaf node with <code>key=00110011</code> corresponds the hash <code>leafHash(33 || 4e07408562bedb8b60ce05c1decfe3ad16b72230967de01f640b7e4729b49fce) = 00be9f2ec46f47e14965f0cb9903f09bc6fe30244109c7c5310180a2251c75cc</code>.*

#### Proof Construction

The properties and protocol for an inclusion proof have been introduced in the Rationale section, above. In this section, we specify how these properties are calculated. We assume that the queried keys have a fixed length `KEY_LENGTH_BYTES`.

We define the following data structures:

```python
@dataclass
class Query:
    key: bytes
    value: bytes
    bitmap: bytes

    def __post_init__(self) -> None:
        # The binary bitmap is the binary string obtained from the bitmap.
        # We will use it also to keep track of the nodes that have been visited
        # and the current height of the query.
        self.binaryBitmap = binaryExpansion(self.bitmap).lstrip("0")

>         self.hash = EMPTY_HASH if self.value == EMPTY_VALUE else LeafNode(self.key, self.value).hash

Author: changed to EMPTY_BYTES

        return queries + [q]

    p = queries[insertIndex]
    if p.binaryPath != q.binaryPath:
        queries.insert(insertIndex, q)

    return queries
```

The following function generates the query response to a single key, including the sibling hashes and the hash of the visited nodes.

```python
def generateExtendedProof(queryKey: bytes, root: bytes) -> ExtendedProof:
    ancestorHashes = []
    siblingHashes = []
    currentNode = getNode(root)
    binaryBitmap = ""

    # Starting from the root, we navigate down the tree.
    height = 0
    while isinstance(currentNode, BranchNode):
        ancestorHashes.append(currentNode.hash)
        if isBitSet(queryKey, height):
            siblingHash = currentNode.leftHash
            currentNode = getNode(currentNode.rightHash)
        else:
            siblingHash = currentNode.rightHash
            currentNode = getNode(currentNode.leftHash)

        if siblingHash == EmptyNode.hash:
            binaryBitmap = "0" + binaryBitmap
        else:
            binaryBitmap = "1" + binaryBitmap
            siblingHashes.append(siblingHash)
        height += 1
    
    # convert binaryBitmap to bytes.

>     bitmap = int(binaryBitmap, 2).to_bytes((len(bstr) + 7) // 8, "big")

Reviewer: Further, note that this will fail for the root where `binaryBitmap` is an emptry string.

>     bitmap = int(binaryBitmap, 2).to_bytes((len(bstr) + 7) // 8, "big")

Author: fixed

```

```python
def smtVerifyInclusionProof(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
    # Check that there is a query for each key.
    if len(queryKeys) != len(proof.queries):
        return False

    for key, query in zip(queryKeys, proof.queries):
        # This is the only difference between an inclusion and a non-inclusion proof.
        if isInclusionProof(key, query) == False:
            return False

    return verify(queryKeys, proof, root)
```

```python
def smtVerifyNonInclusionProof(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
    # Check that there is a query for each key.
    if len(queryKeys) != len(proof.queries):
        return False

    for key, query in zip(queryKeys, proof.queries):
        # This is the only difference between an inclusion and a non-inclusion proof.
        if isInclusionProof(key, query) == True:
            return False
            
    return verify(queryKeys, proof, root)
```

```python
def verify(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
    for key, query in zip(queryKeys, proof.queries):
        # The bitmap does not contain extra leading 0s.
        if query.bitmap[0] == 0:

>             return False

Reviewer: This means that verify on a tree containing just one key-value pair will fail.

Author: in that case I think the proof queries should be empty, as the key value would be the root

Reviewer: Assume, the STM contains only the key value pari `(a,b)`. I then call `generateProof(a, [a]) and obtain
```
Proof:
    siblingHashes = []
    queries = [Query( key = a, value =b, bitmap = 0x00]
```

Author: This is maybe not well defined then. I would imagine that the bitmap should be empty in this case. I will test this again

Author: I get an empty bitmap in the case you describe (and indeed I added an extra check to fix the edge case)

Reviewer: Okay, works now.

        if len(key) != KEY_LENGTH_BYTES:
            return False

        # q is an inclusion proof for key or a default empty node.
        if key == query.key:
            continue

        # q is an inclusion proof for another leaf node.
        commonPrefix = os.path.commonprefix([binaryExpansion(query.key), binaryExpansion(key)])
        if len(query.binaryBitmap) > len(commonPrefix):
            # q does not give an non-inclusion proof for key.
            return False

    filteredQueries = []
    _filter: dict[bytes, Query] = {}
    for q in proof.queries:
        # Remove duplicate queries preserving order.

>         # This can happen if the same query is given for different queryKeys, as for inclusion proofs or non-inclusion proofs pointing to a leaf node

Reviewer: We do not allow mixing inclusion and non-inclusion proofs, right?

Author: The verify function allows it, the new APIs do not allow it

Reviewer: But what `verify` actually work correctly if we mix inclusion and non-inclusion proofs? I thought that was the problem and therefore the new public functions distinguish these cases and verify is an internal/private function that may assumed to be called only for inclusion or non-inclusion proofs.

Author: It works correctly if you intend its usage to just calculate a root from a proof and verify that it matches another root. It does not distinguish whether the proof is an inclusion or non-inclusion proof.

        # or for non-inclusion proofs pointing to the same empty node.
        # To do this, we check the binary path (key binary expansion up to the query height).
        if q.binaryPath not in _filter:
            filteredQueries.append(q)
            _filter[q.binaryPath] = q
        # If the query is a duplicate, check that the two queries are consistent.
        else:
            duplicateQuery = _filter[q.binaryPath]
            # Check that the value and bitmap are the same.
            if q.value != duplicateQuery.value or q.bitmap != duplicateQuery.bitmap:
                return False
            # If the value is empty, both queries point to an empty node.
            # The binary paths must match (the keys can be different).
            if q.value == EMPTY_VALUE:
                if q.binaryPath != duplicateQuery.binaryPath:
                    return False
            # Otherwise, the keys must be the same.
            else:
                if q.key != duplicateQuery.key:

>                     return False

Author: I joined the checks

>                     return False

Reviewer: Maybe the comment could still be updated to
"In case of inclusion proofs, i.e., if the value of q is empty, the keys must be the same".
Do we here implicitly assume to only either exclusively have inclusion or exclusive non-inclusion proofs?

>                     return False

Author: we don't assume that, they can be mixed.

>                     return False

Author: Regarding the comment, in case of inclusion proof the value of q is non empty. Also, if the value of q is non empty, this is not necessarily an inclusion proof

            return False
        if len(key) != KEY_LENGTH_BYTES:
            return False

        # q is an inclusion proof for key or a default empty node.
        if key == query.key:
            continue

        # q is an inclusion proof for another leaf node.
        commonPrefix = os.path.commonprefix([binaryExpansion(query.key), binaryExpansion(key)])
        if len(query.binaryBitmap) > len(commonPrefix):
            # q does not give an non-inclusion proof for key.
            return False

    filteredQueries = []
    _filter: dict[bytes, Query] = {}
    for q in proof.queries:
        # Remove duplicate queries preserving order.
        # This can happen if the same query is given for different queryKeys, as for inclusion proofs or non-inclusion proofs pointing to a leaf node
        # or for non-inclusion proofs pointing to the same empty node.
        # To do this, we check the binary path (key binary expansion up to the query height).
        if q.binaryPath not in _filter:
            filteredQueries.append(q)
            _filter[q.binaryPath] = q
        # If the query is a duplicate, check that the two queries are consistent.
        else:

>             duplicateQuery = _filter[q.binary_path]

Reviewer 2: ```suggestion
            duplicateQuery = _filter[q.binaryPath]
```

            # Check that the value and bitmap are the same.
            if q.value != duplicateQuery.value or q.bitmap != duplicateQuery.bitmap:
                return False
            # If the value is empty, both queries point to an empty node.
            # The binary path must match (the keys can be different).
            if q.value == EMPTY_VALUE:

>                 if q.binary_path != duplicateQuery.binary_path:

Reviewer 2: ```suggestion
                if q.binaryPath != duplicateQuery.binaryPath:
```

        # We will use it also to keep track of the nodes that have been visited
        # and the current height of the query.
        self.binary_bitmap = binary_expansion(self.bitmap).lstrip("0")
    
    @property
    def height(self) -> int:
        return len(self.binary_bitmap)

    @property
    def binary_path(self) -> str:
        # Convert the key to binary string and take the first self.height digits.
        return binary_expansion(self.key)[:self.height]
```

```python
@dataclass
class QueryProof(Query):
    ancestor_hashes: list[bytes]
    sibling_hashes: list[bytes]
```

```python
@dataclass
class Proof:
    sibling_hashes: list[bytes]
    queries: list[Query]
```

We make use of the following utility functions:

```python
def binary_expansion(key: bytes):
    '''
    Convert a key to its binary representation
    '''
    return f"{int.from_bytes(key, 'big'):0{8*len(key)}b}"
```

```python
def are_sibling_queries(q1: Query, q2: Query) -> bool:
    '''
    This function returns true if two queries are sibling,
    i.e. they point to two sibling nodes, false otherwise.
    '''
    if len(q1.binary_bitmap) != len(q2.binary_bitmap):
        return False

>     if q1.binary_key[:q1.height - 1] != q2.binary_key[:q.height - 1]:

Author: I define binaryKey now

        return False

    return q1.binary_key[q1.height - 1] != q.binary_key[q1.height - 1]
```

```python
def insert_query(
    q: Query, queries: list[Query]
) -> list[Query]:
    '''
    This function inserts the a query back in the query list in the correct positon
    to keep the list sorted by longest binary bitmap (bigger height), breaking ties by smaller key.
    This corresponds to a bottom-up left-right ordering of the corresponding leaf nodes.
    We assume that the input queries list is already sorted.
    '''
    if len(queries) == 0:
        return [q]

>     # Run a binary search to efficiently find the insertion index.

Author: I think a binary search is not too technical and the comment should clearly explain what we do. If we remove the binary search we get a O(N^2) routine (N is the number of queries) which we can easily avoid with the binary search.

>     # Run a binary search to efficiently find the insertion index.

Reviewer: I think the specifications should choose the right data structures and algorithms to be efficient, but should otherwise aim for simplicity. We should also use libraries for standard algorithms where possible (e.g., https://docs.python.org/3/library/bisect.html).

Note that the running time of `insertQuery` in Python is O(n) as insertion in a list is O(n). So instead of
```
queryProofs = insertQuery(q, queryProofs)
```
you can do
```
queryProofs = sorted(queryProofs + [q], key=lambda q: (-q.height, q.key))
```
with basically same time complexity as sorting with Timsort (https://en.wikipedia.org/wiki/Timsort) in Python should be O(n) in an almost sorted listed (you could still add a comment to use bisection for optimization in the implementation). Alternatively, I would use the bisect library.

>     # Run a binary search to efficiently find the insertion index.

Reviewer: Okay, let us keep it then. Note the running time of `insertQuery` is O(n) due to insertion being O(n) and also with basic list iteration it would stay O(n) not O(n^2).

>     # Run a binary search to efficiently find the insertion index.

Author: O(n^2) I meant calling insertQuery (O(n)) for each query (O(n)), but we agree then, resolving

    low = -1
    high = len(queries)
    while 1 + low < high:
        middle = low + ((high - low) >> 1)
        p = queries[middle]
        if (q.height == p.height and q.key < p.key) or (q.height > p.height):
            high = middle
        else:
            low = middle

    insert_index = high

    if insert_index == len(queries):
        return queries + [q]

    p = queries[insert_index]

>     if p.binary_path != q.binary_path: # If another query at the same 

Author: This happens when two queries merge. The binary path is the binary expansion up to the query height. Basically when two queries merge we just need one of them

>     if p.binary_path != q.binary_path: # If another query at the same 

Reviewer: Shouldn't I check then that the value and other  properties are also identical?

>     if p.binary_path != q.binary_path: # If another query at the same 

Reviewer: So you use `insertQuery` in `generateProof`. In that case, we modify only `q.binaryBitmap` of each query, but key, value, hash, bitmap stay the same and only `binaryBitmap` gets truncated with each level.

In that case, if the binary path of two queries is the same, `binaryBitmap` (i.e., the witness bitmap) should be the same. However, key, value and bitmap can be different.

The second place you use the function is in `calculateRoot`. Here you actually update the hash value of each query and truncate  `binaryBitmap`. I think here `key`, `value` and `bitmap` can still be different, as they correspond to the original query, but `hash` should coincide and also `binaryBitmap`.

>     if p.binary_path != q.binary_path: # If another query at the same 

Reviewer: Thinking more about it, I think it is correct as is. The reason is basically the following in `calculateRoot`:

```python
# we distinguish three cases for the sibling hash:
        # 1. sibling is next element of sortedQueries
        if len(sortedQueries) > 0 and areSiblingQueries(q, sortedQueries[0]):
            siblingHash = sortedQueries[0].hash
            # We are merging two branches.
            # Check that the bitmap at the merging point is consistent with the nodes type.
            if sortedQueries[0].hash == EMPTY_HASH:
                assert q.binaryBitmap[0] == "0"
            else:
                assert q.binaryBitmap[0] == "1"
            if q.hash == EMPTY_HASH:
                assert sortedQueries[0].binaryBitmap[0] == "0"
            else:
                assert sortedQueries[0].binaryBitmap[0] == "1"

            # Check that the bitmap coincide from the merging point up.
            if len(q.binaryBitmap) > 0:
                assert q.binaryBitmap[1:] == sortedQueries[0].binaryBitmap[1:]

            del sortedQueries[0]
```

So I delete any sibling that is member of sortedQueries before calling `insertQuery`. This means that I should actually always insert the query back in the function `insertQuery` as the following check is always satisfied:
```
  if p.binaryPath != q.binaryPath:
        queries.insert(insertIndex, q)
 ```
